### ISSUE TITLE
[3] Implement batch mode for the search_by_provider API

### ISSUE THREAD

**Issue by hyanwong** [EOL/tramea#117]
As in https://github.com/EOL/tramea/issues/40 but for the search_by_provider API (http://eol.org/api/docs/search_by_provider). This would also help @iimog (and @jhpoelen too, I suspect)

**Comment by iimog** [comment]
This would indeed help me a lot.

**Comment by jhpoelen** [comment]
Great idea! For my purposes, I was hoping for EOL for offer all data in the form of a file archive. see https://github.com/EOL/tramea/issues/10 . File archives can be archived, published and easily embedded in automated high volume data processing workflows. However, I think that the batch API approach is helpful for smaller scale analysis and/or exploratory web apps.

**Comment by jhpoelen** [comment]
Please see https://github.com/jhpoelen/eol-globi-data/issues/145 for an example of a archive based approach.

**Comment by jhammock** [comment]
We'd like to position this for Soon, but double checking the intent of the ticket, got the following:

@hyanwong: Well, I 'really want' to convert Open Tree of Life IDs to EOL page IDs for a batch of ~ 50 leaves on the open tree of life, to add pictures to leaves on the open tree of life. This can be done as high throughput when BioGUID gets off the ground, but until then, I can do it by taking the ncbi ID of an OpenTree species and mapping it to a page ID via the search_by_provider API.

@hyanwong: @jhammock yes, (an archive version of) this would be useful and I think is in hand via @dimus already. But the API route might be useful too, and will be more up-to-date.

@jhammock yes, for my specific task an archive would be sufficient. Thanks!

@jhammock is curious about this existing Global Names archive service; if @dimus could provide a pointer, that would be great. We don't want to duplicate effort, but we can learn from examples; we have other archives we need to make available...

Guessing tentatively that the batch mode a la #40 is still a good and easy solution for now, but standing by for other distinctions that may be important...

**Comment by jhammock** [comment]
Use cases:

@hyanwong 
Thinking of using this for a page with 50 species, each of which may have (say) 3 provider ids

**Comment by hyanwong** [comment]
At the moment the API is called using e.g. http://eol.org/api/search_by_provider/1.0/123?hierarchy_id=903, which is not condusive to passing in multiple taxon IDs like 123, 456, 789, ... for a single hierarchy_id. Without breaking the current API it might be sensible to add an alternative notation to allow the first number to be interpreted as a hierarchy_id instead, if (for example) the query string contains 'id=' rather than 'hierarchy_id='. E.g. to refer to taxon 123 in hierarchy_id 903, you could use

http://eol.org/api/search_by_provider/1.0/903?id=123

Then it would be easy to allow 

http://eol.org/api/search_by_provider/1.0/903?id=123&id=456&id=789

or, I guess, separated by commas (%2C when uri-escaped)

http://eol.org/api/search_by_provider/1.0/903?id=123%2C456%2C789

**Comment by JRice** [comment]
[IGNORE THIS COMMENT]

I'm just writing this here to get some thoughts out for the next version of EOL, because I hate this API.

Why do I hate this API?

For one, it's poorly named; no where else but in the API do we refer to "providers," we have "partners," and even then we actually talk about these identifiers in terms of _resources_, not partners. (And see later comments for discussion around the word "search".) And the ID for the "provider" comes from a completely differently-named API ("provider_hierarchies"), etc, etc... it's not even clear (to me) how I get from a NAME of a provider to the ID we're using. Thus, some thoughts on where we could make this cleaner and more aligned with everything else:

It's fast to query on the partner's identifier. For example,

```
select * from hierarchy_entries where identifier = '123';
83 rows in set (0.19 sec)
```

It's not much slower to query on several (it seems to scale more or less linearly, at least with small numbers).

```
select * from hierarchy_entries where identifier IN ('123','456','789');
215 rows in set (0.24 sec)
```

I could imagine an interface where the user says "I have this ID" and we say "yup, we see that ID is used by CoL (to this page), ITIS (to this page) and blahblablha (to this page)" by calling, if you'll bear with my name changes for a moment because I'm thinking of Tramea, `eol.org/nodes.json?resource_pk=123` Because I'm thinking about more generic API interfaces, I actually imagine this returning pretty robust results (I am lazily NOT quoting left-hand identifiers, which is less-than-clean JSON, sorry):

```
{
total: 83,
page: 1,
per_page: 50,
nodes: [
  { id: 487400,
    resource: { name: "CoL", id: 1},
    parent_id: 1232,
    root_id: 1,
    rank: { id: 1, name: "species" },
    original_rank: { id: 123, name: "sp." },
    scientific_name: "Foo baricus Posford 1922",
    canonical_form: "Foo baricus",
    remarks: "",
    resource_pk: "123",
    page: { id: 23422, image: "blahblah", etc; there will be a standard "packet" of info about a page }
  },
  { ...etc: 49 more results here  ... }
]
}
```

Assuming API callers might want a "light" view, we might want to add an option such as `eol.org/nodes.json?resource_pk=123&view_style=identifiers`, which might just return a node like `{ id: 487400, resource_id: 1, parent_id: 1232, root_id: 1, rank_id: 1, resource_pk: "123", page_id: 23422 }`.

One could then imagine this being filtered by calling, say, `eol.org/nodes.json?resource_pk=123&resource_id=2`, which would return a single result (or none).

And, of course, "batches" of IDs as Yan suggests: `eol.org/nodes.json?resource_pk=123&resource_pk=234&resource_pk=456&resource_id=2`, which would return 0-3 results.

...This approach generalizes things a bit, associates the API with what one could easily turn into a web UI (though I suspect it would not get use—other than perhaps API users who want to see a rendered result—and we would not link to it), and makes things less "proprietary" in the sense of "we tacked on this API call because we needed the functionality" and more "we added this API because it jives with the structure of our data... and happens to be useful."

I still think MOST people consuming what is now our search_by_provider API are going to want a DB dump, since (presumedly) they have a bunch of IDs and just want a matching set of other people's IDs. ...but this would make sense as an API and serve that function when needed.

I could be wrong. Just thinking out loud.

**Comment by hyanwong** [comment]
Agree about naming and syntax. Haowever, I can't see many people having an ID and not knowing where it is from, so some sort of indicator of "this id is from CoL" or "this id is from NCBI" is presumably the norm.

I actually have "this single taxon is represented in my db by ncbi=X and gbif=Y - give me the best single EoL page match", but that's a specific use case.

I don't know if "most" will want a DB dump. Imagine (say) ncbi wanting a picture to decorate their page for a taxon. Will they want to do it for all pages at once with a DB dump, or every time someone visits the page, do they want to get the best recent picture available via an AJAX call to the EOL API? 50:50 I would say.

**Comment by JRice** [comment]
> I can't see many people having an ID and not knowing where it is from, so some sort of indicator of "this id is from CoL" or "this id is from NCBI" is presumably the norm.

Absolutely. I was trying to point out that simple filters would allow the functionality and alleviate the confusion about what goes where (i.e.: "if they give us an explicit 'id' parameter, assume the implicit ID in the URL is the provider id; if they give us an explicit provider id, assume the implicit ID is an identifier", etc, etc... clear as mud. :)

And point taken about who wants what. I'm not an API consumer. My opinion is not to be trusted. :)

**Comment by hyanwong** [comment]
Yes. The main point is that it's clunky trying to retain backwards compatibility. Another possibility would be to bump the API to v 2.0 and change the syntax

http://eol.org/api/search_by_provider/2.0/903?id=123

And not allow "hierarchy_id=X" at all, but require it as the first number of the URL (not in the parameters). That makes a lot more sense to me.

**Comment by hyanwong** [comment]
Or indeed keep "search_by_provider", but mark as obsolete, and implement "search_by_partner" as the name for its replacement

http://eol.org/api/search_by_partner/1.0/903?id=123

or whatever word you want to substitute for "provider".

**Comment by JRice** [comment]
Yup. There's resistance to bumping any API versions ('cause one would naturally expect them all to get bumped, and if they all get bumped, there's lots of stuff we want to add). ...probably not going to happen. :\ Probably much easier to add a new API call under 1.0.

...Also, why is this called "search"? It's not really a _search_, any more than eol.org/pages/1234 is a "search" for page 1234. You have an id, you're retrieving information about it. I don't get the name.

I suppose if I _had_ to pick a new name for this beast, I would choose `source_identifiers`, because that's what these things are called in the existing `hierarchy_entries` API. It's still lame, but at least it's consistently lame. I wouldn't force you to pass in EITHER value, since we don't know which one you have. So you would be explicit in both cases: `eol.org/api/source_identifiers/1.0?hierarchy_id=23&identifier=123`. But I'm a jerk.  ;)

**Comment by hyanwong** [comment]
'one would naturally expect them all to get bumped' - really? I wouldn't. But that's just me.

Not sure I agree about not requiring a value for hierarchy_id. I really can't imagine any likely use cases. But I could see a batch case where you might want different hierarchy_ids:

hierarchy_plus_id=23+123&hierarchy_plus_id=903+567, etc. But that's probably too elaborate. Better in that case to require a separate API call for each hierarchy_id, I think. There aren't that many of them, after all.

**Comment by JRice** [comment]
You are probably right (Re: required hierarchy_id). I'm just not sure it naturally "fits" anywhere like that. You have one hierarchy id and one or more identifier; you want back page ids (or "taxon concept ids" or "eol_page_ids" or whatever). Hmmn...

In the current API, if you have...
- a page id and want to get more detail about it, you call **pages**.
- a name and you want a page id for it, you call **search**.
- nothing, and want a list of hierarchy ids, you call **provider_hierarchies**.
- a hierarchy id, and you want ...a list of root nodes (?!?), you call **hierarchies**.

...It all makes ... no sense. At all. It's not at all aligned. Sigh. I suppose "search_by_provider" is closest to "I have an id and I want more detail about it", even though the "detail" is pretty limited. So it should be named based on what you have, and what you have is a hierarchy's identifier.

`eol.org/hierarchy_identifiers/[hierarchy_id]?identifier=[some text, usually numeric]` ?

I could agree to that. I don't like it, but it _kinda_ fits

**Comment by hyanwong** [comment]
I guess you also include the 5th bullet point. You have:
- a hierarchy id and one or more identifier strings in that hierarchy, and you want ...a list of EOL page IDs, you call <whatever_the_search_by_providerAPI_is_now_called>.

That returns EoL pages ides. Then to get detail about those pages, you have to do a further call to the Pages API. Kinda makes sense.

**Comment by hyanwong** [comment]
Oh, and I think it is a search of some form, because the provider ID might not even exist in the EoL database, for some reason. You are 'searching' for the EOL page that corresponds to a taxon in some other provider's humungous list of taxa.

**Comment by JRice** [comment]
Sure, in the same way eol.org/pages/1234 is a search; it might not be there (or might not be visible to you).

In Rails parlance, this is called a "find," not a search. I am likely splitting hairs, but I regardless, don't like the term "search" here, and would prefer "hierarchy_identifiers" to more closely align (kinda) with the other APIs (which themselves are not in great alignment).

Probably not worth debating further. I don't actually care what we do to meet the immediate need: whatever works. I'm much more interested in how we handle this in Tramea.  ;)

**Comment by hyanwong** [comment]
'whatever works' agreed. Adding comments here simply to preserve ideas / use cases for Tramea

**Comment by JRice** [comment]
I feel compelled to add: how search_by_provider is currently implemented is obfuscating what's happening behind the scenes, and gives the wrong impression of how EOL works. What's actually happening is that you're looking up a hierarchy_entry using the "identifier" field (and the hierarchy_id field) instead of the "id" field. This isn't a new feature, it's not associated with a new table, it's not a search (which IMO involves some form of "fuzziness" or another), but a lookup. ...What I proposed for Tramea could actually be implemented on the hierarchy_entries API, though not w/o major changes, so I don't recommend it.

Maybe I'm starting to repeat myself.  :S

**Comment by jhpoelen** [comment]
For what it is worth (I only glanced over this impressive thread): I like the idea of asking eol for a page for [WoRMS taxon 105838](http://www.marinespecies.org/aphia.php?p=taxdetails&id=105838). It makes for easy linking and I don't have to worry about some magic ids associated with the internal representation of WoRMS taxonomy. 

I took a similar approach with GloBI: WoRMS uses their own ids to link into GloBI using
http://www.globalbioticinteractions.org/?sourceTaxon=WORMS%3A105838&interactionType=interactsWith .

**Comment by jhammock** [comment]
This should be deployed, I think. Is it working?

**Comment by jhammock** [comment]
@HudaSoliman will check whether this is merged, and/or what happened
